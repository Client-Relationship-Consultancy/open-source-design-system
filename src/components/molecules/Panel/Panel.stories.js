import React from "react"
import { storiesOf } from "@storybook/react"

import Panel from "."

const stories = storiesOf("Panels", module)

stories.add("Show/Hide Panel", () => {
    return (
        <Panel defaultOpen label="Toggle Panel" icon="graduation-cap">
            <h3>This content can be hidden by clicking the button.</h3>
            <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse orci ex, 
                feugiat vitae augue at, ultricies maximus dui. Class aptent taciti sociosqu ad 
                litora torquent per conubia nostra, per inceptos himenaeos. Sed elementum lectus nulla, 
                at placerat felis interdum ut. Etiam vel eros feugiat, egestas ante eu, hendrerit nibh. 
                ivamus lacus dui, tempus et risus ac, congue semper mi. Aenean ut viverra tortor, ac semper diam. 
                Curabitur sit amet venenatis lacus. In facilisis eros ac pharetra faucibus. Duis non nisi neque. 
                Curabitur vitae porta mauris, in sollicitudin ex. Nullam ut risus risus. </p>
        </Panel>
    );
})
