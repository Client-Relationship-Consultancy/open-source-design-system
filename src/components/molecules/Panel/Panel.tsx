import React from "react";
import styled from "styled-components"
import { IconProp } from "@fortawesome/fontawesome-svg-core";
import { colourPalette } from "../../../brandColours"

import Icon from "../../atoms/Icon"

export const PanelButton = styled.button`
display: flex;
  flex-direction: row;
  align-items: center;
  height: 1.5rem;
  padding: 0;
  margin: 0.5rem 0;
  text-decoration: none;
  border: none;
  font-size: 1rem;
  background-color: transparent;
  color: ${props => props.theme.action.main.hex};
  path {
    fill: ${props => props.theme.action.main.hex};
  }
  :hover {
    cursor: pointer;
  }
  :focus {
      outline: none;
  }
`
PanelButton.defaultProps = {
    theme: colourPalette.examplePalette,
  }

PanelButton.displayName = "PanelButton"

const StyledIcon = styled(Icon)`
transform: scale(0.8, 1.25);
    margin-right: 0.5rem;
`
StyledIcon.displayName = "StyledIcon"

interface IArrowIconProps {
    flipped: boolean;
}

export const ArrowIcon = styled(Icon)<IArrowIconProps>`
    margin-left: 0.5rem;
    padding-top: 0.1rem;
    svg {
        transform: scaleY(${(props) => props.flipped ? -1 : 1});
        transition: transform 0.3s ease-in-out;
    }
`
ArrowIcon.displayName = "ArrowIcon"

interface IPanelProps {
    defaultOpen: boolean;
    icon?: IconProp;
    label: string | React.ReactNode; 
    children: React.ReactNode;
}

interface IPanelState {
    open: boolean;
}

class Panel extends React.PureComponent<IPanelProps, IPanelState> {
    state: IPanelState = {
        open: this.props.defaultOpen,
    }

    toggleOpen = () => {
        this.setState(prevState => ({ 
            open: !prevState.open, 
        }));
    }

    render(){
        return (
            <>
                <PanelButton onClick={this.toggleOpen}>
                    {this.props.icon ? <StyledIcon name={this.props.icon} size="1x" /> : null }
                    {this.props.label}
                    <ArrowIcon name="chevron-down" size="1x" flipped={this.state.open} />
                </PanelButton>
                {this.state.open && this.props.children}
            </>
        );
    }
}

export default Panel;
