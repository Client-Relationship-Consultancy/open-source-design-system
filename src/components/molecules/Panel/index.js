import Panel from "./Panel"

export { ArrowIcon } from "./Panel"
export default Panel
