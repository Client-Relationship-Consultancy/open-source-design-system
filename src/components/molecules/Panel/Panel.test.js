import React from "react"
import { shallow } from "enzyme"
import toJson from "enzyme-to-json"
import "jest-styled-components"
import Panel from "./Panel"

describe("Panel", () => {
    const commonProps = {
        defaultOpen: false,
        icon: "check",
        label: "open/close",
    }
    it("should match the snapshot", () => {
        const componentWithoutChild = shallow(<Panel {...commonProps}/>)
        const tree = toJson(componentWithoutChild)
        expect(tree).toMatchSnapshot()
    })
    it("should show child content when the button is clicked", () => {
        const childProp = <p>Mock Content</p>;
        const componentWithChild = shallow(<Panel {...commonProps}>{childProp}</Panel>)
        expect(componentWithChild.state().open).toEqual(false)
        expect(componentWithChild.find("p").exists()).toEqual(false)
        componentWithChild.find("PanelButton").simulate("click")
        expect(componentWithChild.state().open).toEqual(true)
        expect(componentWithChild.find("p").exists()).toEqual(true)
    })
    // it("should use the correct colours", () => {
    //     const childProp = <p>Mock Content</p>;
    //     const componentWithChild = shallow(<Panel {...commonProps}>{childProp}</Panel>)
    //     console.log(componentWithChild.find("PanelButton").props())
    //     expect(componentWithChild.find("PanelButton")).toHaveStyleRule("color", "#01579B")
    // })
    // it("should flip the arrow icon when button is clicked", () => {
    //     const componentWithoutChild = shallow(<Panel {...commonProps}/>)
    //     // console.log(componentWithoutChild.find("ArrowIcon").children().svg.debug())
    //     // expect(componentWithoutChild.find("ArrowIcon")).toHaveStyleRule("transform", "scaleY(1)")
    //     expect(componentWithoutChild.find("ArrowIcon").toHaveStyleRule("margin-left", "0.5rem")
    //     componentWithoutChild.find("PanelButton").simulate("click")
    //     // expect(componentWithoutChild.find("ArrowIcon")).toHaveStyleRule("transform", "scaleY(-1)")
    // })
})

