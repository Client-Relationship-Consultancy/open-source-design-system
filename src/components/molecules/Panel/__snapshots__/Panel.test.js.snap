// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel Component Testing should match the snapshot 1`] = `
<Fragment>
  <PanelButton
    onClick={[Function]}
  >
    <StyledIcon
      name="check"
      size="1x"
    />
    open/close
    <ArrowIcon
      flipped={false}
      name="chevron-down"
      size="1x"
    />
  </PanelButton>
</Fragment>
`;

exports[`Panel should match the snapshot 1`] = `
<Fragment>
  <PanelButton
    onClick={[Function]}
  >
    <StyledIcon
      name="check"
      size="1x"
    />
    open/close
    <ArrowIcon
      flipped={false}
      name="chevron-down"
      size="1x"
    />
  </PanelButton>
</Fragment>
`;
